scanner/revs.go

```diff
@@ -22,139 +21,194 @@ const (
 
-// RevListShas uses git rev-list to return the list of object sha1s
-// for the given ref. If all is true, ref is ignored. It returns a
-// channel from which sha1 strings can be read.
+	// ambiguousRegex is a regular expression used to determine whether or
+	// not a call to `git rev-list ...` encountered ambiguous refs.
+	ambiguousRegex = regexp.MustCompile(`warning: refname (.*) is ambiguous`)
+)
+
+// RevListShas list all revisions between `refLeft` and `refRight` according to
+// the rules provided by *ScanRefsOptions.
 func RevListShas(refLeft, refRight string, opt *ScanRefsOptions) (<-chan string, <-chan error, error) {
-	var stdin []string
+	return NewRevListScanner(opt).Scan(refLeft, refRight)
+}
 
-	refArgs := []string{"rev-list", "--objects"}
+// RevListScanner implements the behavior of scanning over a list of revisions.
+type RevListScanner struct {
+	// ScanMode is an optional parameter which determines how the
+	// RevListScanner should process the revisions given.
+	ScanMode ScanningMode
+	// RemoteName is the name of the remote to compare to.
+	RemoteName string
+	// SkipDeletedBlobs provides an option specifying whether or not we
+	// should report revs whos blobs no longer exist at the remote, or at
+	// the range's end.
+	SkipDeletedBlobs bool
+
+	revCache *RevCache
+}
 
-	switch opt.ScanMode {
-	case ScanRefsMode:
-		if opt.SkipDeletedBlobs {
-			refArgs = append(refArgs, "--no-walk")
-		} else {
-			refArgs = append(refArgs, "--do-walk")
-		}
+// NewRevListScanner constructs a *RevListScanner using the given opts.
+func NewRevListScanner(opts *ScanRefsOptions) *RevListScanner {
+	return &RevListScanner{
+		ScanMode:         opts.ScanMode,
+		RemoteName:       opts.RemoteName,
+		SkipDeletedBlobs: opts.SkipDeletedBlobs,
 
-		refArgs = append(refArgs, refLeft)
-		if refRight != "" && !z40.MatchString(refRight) {
-			refArgs = append(refArgs, refRight)
-		}
-	case ScanAllMode:
-		refArgs = append(refArgs, "--all")
-	case ScanLeftToRemoteMode:
-		args, commits := revListArgsRefVsRemote(refLeft, opt.RemoteName)
-		refArgs = append(refArgs, args...)
-		if len(commits) > 0 {
-			stdin = commits
-		}
-	default:
-		return nil, nil, errors.New("scanner: unknown scan type: " + strconv.Itoa(int(opt.ScanMode)))
+		// TODO(taylor): fix dependency on having mutable data in "opts"
+		revCache: NewRevCache(opts.nameMap),
 	}
+}
 
-	// Use "--" at the end of the command to disambiguate arguments as refs,
-	// so Git doesn't complain about ambiguity if you happen to also have a
-	// file named "master".
-	refArgs = append(refArgs, "--")
+// Scan reports a channel of revisions, and a channel of errors encountered
+// while scanning those revisions, over the given range from "left" to "right".
+//
+// If there was an error encountered before beginning to scan, it will be
+// returned by itself, with two nil channels.
+func (r *RevListScanner) Scan(left, right string) (<-chan string, <-chan error, error) {
```

Comment:
I was debating whether or not to use the *XChannelWrapper` types when I was refactoring this. IMHO, I think passing the channels around directly and using the `tools.CollectErrsFromChan` method is more lightweight than a new type, so I am going to leave this as-is for now.

This will become less of a sore spot a few more PRs down the line.